### ISSUE TITLE
Version tag specification

### ISSUE THREAD

**Issue by HaleTom** [robert7/nixnote2#28]
I am working on packaging this for Arch Linux's AUR.

Currently I see that development version is obtained by:

````
head ./debian/changelog --lines=1|sed -E 's/^[^(]*\(([^)]*).*/\1/g'
````

What I'm trying to work out is which commit to use for packaging.

How do you currently specify versions?

Would it be fair conclude that I could obtain versions as follows:

**Non-alpha release:**

```
$ git tag --list | grep '^v' | grep -v alpha | tail -n1
v2.1.0-beta-3
```

**Full release:**

```
$ git tag --list | grep '^v' | grep -Ev 'alpha|beta' | tail -n1
v2.0.2
```

Would you mind writing out the specification for your version numbering?  Ideally, it would answer the following questions:

1. What is the difference between the `continuous` and `continuous-develop` branches?
1. Besides these two, do you only have `alpha`, `beta` and full releases?
1. Will there continue to be a `-` between `(alpha|beta)` and the number? eg: `v2.1.0-beta-1`
1. What is the meaning of the final letter `a` in `v2.1.0-beta-4a`?
1. Will unannotated git tags continue to describe only full releases?

1. Will my commands above accurately capture your release tagging going forward?

**Comment by HaleTom** [comment]
Thanks!

I just noticed:

`v2.1.0-beta-1` has 2 `-` but `v2.1.0-alpha10` only has one. 

You'll be going with two?

**Comment by robert7** [comment]
So I created a wiki page with description: [Releases versions, build pipeline, branches, tags...](https://github.com/robert7/nixnote2/wiki/Releases---versions,-build-pipeline,-branches,-tags)
Pls look at it if it answer all question. This is first version - I will improve it later.

There were 2 hyphens used pre-release tags e.g. in "v2.1.0-beta-1", but I would in the future only use one hyphen. So it will be like "v2.1.0-beta1". Version inside application and in the "debian/changelog" file is without the prefix "v". The prefix "v" follows the naming convention of tags from upstream.

Further you can distinguish the "pre release" from "release" according to semantic versioning by checking of the presence of the hyphen character. This is better then filtering out "alpha and beta" words. So you may adjust the "grep" examples. I copied them into the wiki page - i'll adjust them later,

I did not understand the question 5:

> "Will unannotated git tags continue to describe only full releases?"

Basically "v" prefix means release. Exceptional/special tags are "continuous-develop" (tip of the development branch) and "continuous" (tip of the master branch). Those are moved after each merge and created as part of the Travis CI pipeline.

**Comment by robert7** [comment]
So tags belonging to "stable releases" are those with prefix "v" and which don't contain hyphen character.

**Comment by HaleTom** [comment]
Cheers for specifying all that! Nice one.

`man git-tag` says:

> Annotated tags are meant for release while lightweight tags are meant for private or temporary object labels. For this reason, some git commands for naming objects (like git describe) will ignore         
       lightweight tags by default.   

```
% git describe --abbrev=0    
v2.0.2
```

Will this continue when you make your first release?

Thinking of the future, what is the process for preparing for a release?

* Is there a new feature freeze? If so, is there a period of time or number of `develop` releases?
* Does the naming reflect intent-to-release, eg `-rc#`?  (Some packagers may appreciate this, as it's likely to be more stable than a `-beta`.)

I look forward to a release soon - you've done bunches of great work. Congratulations and thank you.

**Comment by HaleTom** [comment]
BTW, your quoting of `'-'` doesn't actually do anything. I think what you're looking for is:
```
git tag --list | grep '^v' | grep -v -- - | tail -n1
```

**Comment by robert7** [comment]
>BTW, your quoting of '-' doesn't actually do anything.

Funny. I tried and it seems both return the same output (at least on my system). But anyway. Just filter out the names with hyphen.

![screen_20181005_01](https://user-images.githubusercontent.com/1135218/46525423-d5eeb380-c88b-11e8-8d1a-a70f3eee5abc.png)

>Annotated tags are meant for release while lightweight tags are meant for private or temporary object 

OK. Actually I did not know that. But it seems reasonable. So I can follow that (i'll update the wiki page later).

>Thinking of the future, what is the process for preparing for a release?
>Is there a new feature freeze?

Ok. The project is quite small. So far with a "team" consisting of "me" :) .. so I didn't want to create "over complicated release process" .. I did no  "rc" releases. But basically the last "beta3" is already a kind of "rc1"... Now the plan is to fix few more minor bugs (I know about few), then do a release. So now its already a kind of feature freeze.

**Comment by HaleTom** [comment]
Awesome! :)

I'll be making 2 "build-from-source" packages:

* The latest `v*` tag, (after #31 is resolved)
* A "release" version (when a new non-hyphenated version is released)

I may also look at making "binary" packages by pulling apart the AppImage for faster package installs (since the build takes some time on my laptop). The bonus of this over the AppImage would be access to documentation.

**Comment by robert7** [comment]
I recreated the tags I created for v2.1 beta versions. I know, this is probably not a "best practice", but its "one time cleanup" and it will make easier in the future to stick with the declared naming/annotation conventions...

1. All release tags have "annotation"
2. Pre-releases contains "hyphen", Release don't contain hyphen
3. Non-release tags (like "continuous") don't have annotation
4. "2 hyphens" in version tag are not used anymore
I'm not sure, if the tags in your repository clone will update with simple "pull", you may need to do a fresh clone.
5. Next release tag will be "v2.1.0-rc1"

It now looks like:
![screen_20181005_01](https://user-images.githubusercontent.com/1135218/46550636-f04d7f00-c8d5-11e8-86c4-00b5b5f7ef30.png)

**Comment by HaleTom** [comment]
Nice alignment.

For other packagers, the command to get the latest versioned tag is:

```
git -c 'versionsort.suffix=-' ls-remote -t --exit-code --refs --sort=-v:refname https://github.com/robert7/nixnote2.git 'v*' | sed -E 's/^[[:xdigit:]]+[[:space:]]+refs\/tags\/(.+)/\1/g'
```

**Comment by HaleTom** [comment]
FYI, I dug deeper into the Arch packaging guidelines, and `-git` packages need to be be built from the HEAD of the default branch of the repository.

(The implied assumption is that the default branch is the "use-this-release" branch)

**Comment by robert7** [comment]
>..packages need to be be built from the HEAD of the default branch of the repository.

So this would be the master branch. this is actually what I would also recommend. 
This is the "latest stable build" (this is also mirrored by the current PPA  build (nixnote-stable) and by the "stable AppImage" build - https://github.com/robert7/nixnote2/releases/tag/continuous
